Fix/meaningfull error message on type update

ulferts opened the pull request:
Fixes https://community.openproject.org/work_packages/21797 and refactors the settings#types. 

Because `project.type_ids=` is used to update the type_ids, it is no longer possible for an invalid state of the project (e.g. a required custom field not filled out) to block the update of the types.

With the introduction of the service, the controller now has way less clutter. I still dislike the heavy usage of flash messages within the controller, especially to transmit error messages, but refactoring that is to big an effort.

ulferts committed:
quickfix for showing a meaningfull error message

And no empty notice.

ulferts committed:
remove unnecessary extend call

ulferts committed:
refactor types action on projects controller

* Introduced a service to reduce complexity in the controller
* Used permitted params to ensure params syntax
* Introduced a scope for querying for types used by a work package of the project

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://ulferts-fix-meaningfull-error-message-on-type-update.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://ulferts-fix-meaningfull-error-message-on-type-update.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

why-el commented on app/services/update_projects_types_service.rb:
```diff
@@ -0,0 +1,59 @@
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+UpdateProjectsTypesService = Struct.new :project do
```
Class syntax here?

oliverguenther replied:
Not entirely sure where this originated, but quite a number of services that use the `Struct` notation and I've come to like the distinction of it, since you almost always need a class with one accessor, a struct does that in one step.

I find it more concise and it internally does nothing but return a new class instance. It only starts to feel weird when you start working with any class methods or initializer here.

why-el replied:
Nope, this usage breaks Inheritance (try it), allows for things like `Service.new` without project even though it is necessary, and ends up including unnecessary modules such as enumerable (and adds weird things to the ancestor chain).

oliverguenther replied:
It obviously breaks inheritance since it implicitly returns an instance. It also does not add weird things to the ancestor chain , it just extends Enumerable to access all members of the struct. 

``` ruby
x = Struct.new :foo
irb(main):012:0> x.ancestors
=> [#<Class:0x007ff3ec3cee38>, Struct, Enumerable, Object, Kernel, BasicObject]

class Foo; end
irb(main):013:0> Foo.ancestors
=> [Foo, Object, Kernel, BasicObject]
```

I agree with the required arguments case, though.

why-el replied:
It doubled the number of ancestors!! :D, but yeah not terribly important. 

Jens agreed he will put in a regular class, since as this grows we might have to maintain a service taxonomy and we need inheritance for that. :+1:

ulferts committed:
use class instead of struct

ulferts commented:
@why-el, @oliverguenther I changed it to be a class and introduced a base class while I was at it. So no more boilerplate for the one writing a project service after me.

oliverguenther commented:
@ulferts Thanks :+1: I've noted a todo for myself to refactor the other services in a dull night.

why-el commented:
Great :)

why-el commented:
We good to merge here?

ulferts commented:
I think so.

oliverguenther commented on app/models/permitted_params.rb:
```diff
@@ -153,6 +153,10 @@ def project_type
   end
 
+  def projects_type_ids
+    params.require(:project).require(:type_ids).map(&:to_i).select { |x| x > 0 }
```
It really doesn't matter here at all, but everytime I see `map+select` I'm reminded that `map+compact` will build no intermediary arrays and [performs better](http://stackoverflow.com/a/5046057).

oliverguenther commented:
I could not reproduce the duplicate flash on dev (only saw `en.timelines.cannot_update_planning_element_types` on type update with a required, missing CF on the project).

You have a point regarding the heavy use of (success) notifications, but that is true for a large potion of the frontend and should be discussed separately.

LGTM. :+1:
